### ISSUE TITLE
VARCHAR requires a length on dialect mariadb

### ISSUE THREAD

**Issue by justinkumpe** [rseichter/fangfrisch#46]
When trying to run initdb I get the following error

```python
Traceback (most recent call last):
  File "/var/lib/fangfrisch/venv/lib/python3.9/site-packages/sqlalchemy/sql/compiler.py", line 6647, in visit_create_table
    processed = self.process(
  File "/var/lib/fangfrisch/venv/lib/python3.9/site-packages/sqlalchemy/sql/compiler.py", line 915, in process
    return obj._compiler_dispatch(self, **kwargs)
  File "/var/lib/fangfrisch/venv/lib/python3.9/site-packages/sqlalchemy/sql/visitors.py", line 141, in _compiler_dispatch
    return meth(self, **kw)  # type: ignore  # noqa: E501
  File "/var/lib/fangfrisch/venv/lib/python3.9/site-packages/sqlalchemy/sql/compiler.py", line 6678, in visit_create_column
    text = self.get_column_specification(column, first_pk=first_pk)
  File "/var/lib/fangfrisch/venv/lib/python3.9/site-packages/sqlalchemy/dialects/mysql/base.py", line 1809, in get_column_specification
    self.dialect.type_compiler_instance.process(
  File "/var/lib/fangfrisch/venv/lib/python3.9/site-packages/sqlalchemy/sql/compiler.py", line 960, in process
    return type_._compiler_dispatch(self, **kw)
  File "/var/lib/fangfrisch/venv/lib/python3.9/site-packages/sqlalchemy/sql/visitors.py", line 141, in _compiler_dispatch
    return meth(self, **kw)  # type: ignore  # noqa: E501
  File "/var/lib/fangfrisch/venv/lib/python3.9/site-packages/sqlalchemy/sql/compiler.py", line 7238, in visit_string
    return self.visit_VARCHAR(type_, **kw)
  File "/var/lib/fangfrisch/venv/lib/python3.9/site-packages/sqlalchemy/dialects/mysql/base.py", line 2309, in visit_VARCHAR
    raise exc.CompileError(
sqlalchemy.exc.CompileError: VARCHAR requires a length on dialect mariadb

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/var/lib/fangfrisch/venv/bin/fangfrisch", line 8, in <module>
    sys.exit(main())
  File "/var/lib/fangfrisch/venv/lib/python3.9/site-packages/fangfrisch/__main__.py", line 63, in main
    DbMeta().create_metadata(args.force)
  File "/var/lib/fangfrisch/venv/lib/python3.9/site-packages/fangfrisch/db.py", line 91, in create_metadata
    DbMeta.init(create_all=True, drop_all=force)
  File "/var/lib/fangfrisch/venv/lib/python3.9/site-packages/fangfrisch/db.py", line 72, in init
    cls.metadata.create_all(cls._engine)
  File "/var/lib/fangfrisch/venv/lib/python3.9/site-packages/sqlalchemy/sql/schema.py", line 5866, in create_all
    bind._run_ddl_visitor(
  File "/var/lib/fangfrisch/venv/lib/python3.9/site-packages/sqlalchemy/engine/base.py", line 3253, in _run_ddl_visitor
    conn._run_ddl_visitor(visitorcallable, element, **kwargs)
  File "/var/lib/fangfrisch/venv/lib/python3.9/site-packages/sqlalchemy/engine/base.py", line 2459, in _run_ddl_visitor
    visitorcallable(self.dialect, self, **kwargs).traverse_single(element)
  File "/var/lib/fangfrisch/venv/lib/python3.9/site-packages/sqlalchemy/sql/visitors.py", line 664, in traverse_single
    return meth(obj, **kw)
  File "/var/lib/fangfrisch/venv/lib/python3.9/site-packages/sqlalchemy/sql/ddl.py", line 918, in visit_metadata
    self.traverse_single(
  File "/var/lib/fangfrisch/venv/lib/python3.9/site-packages/sqlalchemy/sql/visitors.py", line 664, in traverse_single
    return meth(obj, **kw)
  File "/var/lib/fangfrisch/venv/lib/python3.9/site-packages/sqlalchemy/sql/ddl.py", line 951, in visit_table
    CreateTable(
  File "/var/lib/fangfrisch/venv/lib/python3.9/site-packages/sqlalchemy/sql/ddl.py", line 314, in _invoke_with
    return bind.execute(self)
  File "/var/lib/fangfrisch/venv/lib/python3.9/site-packages/sqlalchemy/engine/base.py", line 1418, in execute
    return meth(
  File "/var/lib/fangfrisch/venv/lib/python3.9/site-packages/sqlalchemy/sql/ddl.py", line 180, in _execute_on_connection
    return connection._execute_ddl(
  File "/var/lib/fangfrisch/venv/lib/python3.9/site-packages/sqlalchemy/engine/base.py", line 1526, in _execute_ddl
    compiled = ddl.compile(
  File "/var/lib/fangfrisch/venv/lib/python3.9/site-packages/sqlalchemy/sql/elements.py", line 308, in compile
    return self._compiler(dialect, **kw)
  File "/var/lib/fangfrisch/venv/lib/python3.9/site-packages/sqlalchemy/sql/ddl.py", line 69, in _compiler
    return dialect.ddl_compiler(dialect, self, **kw)
  File "/var/lib/fangfrisch/venv/lib/python3.9/site-packages/sqlalchemy/sql/compiler.py", line 870, in __init__
    self.string = self.process(self.statement, **compile_kwargs)
  File "/var/lib/fangfrisch/venv/lib/python3.9/site-packages/sqlalchemy/sql/compiler.py", line 915, in process
    return obj._compiler_dispatch(self, **kwargs)
  File "/var/lib/fangfrisch/venv/lib/python3.9/site-packages/sqlalchemy/sql/visitors.py", line 141, in _compiler_dispatch
    return meth(self, **kw)  # type: ignore  # noqa: E501
  File "/var/lib/fangfrisch/venv/lib/python3.9/site-packages/sqlalchemy/sql/compiler.py", line 6657, in visit_create_table
    raise exc.CompileError(
sqlalchemy.exc.CompileError: (in table 'refreshlog', column 'url'): VARCHAR requires a length on dialect mariadb
```

This is my config file:
```python
[DEFAULT]
db_url = mariadb+pymysql://[username]:[password]@[server]/[database]?charset=utf8mb4
local_directory = /var/lib/clamav
max_size = 5MB
on_update_exec = clamdscan --reload
on_update_timeout = 42

[interserver]
enabled = yes

[sanesecurity]
enabled = yes
# Use a non-default mirror
prefix = https://mirror.rollernet.us/sanesecurity/

[urlhaus]
enabled = yes
max_size = 2MB
```

**Comment by rseichter** [comment]
That doesn't look like something I can help you with. You probably have a better chance of getting useful information from the SQLAlchemy developers. Fangfrisch delegates all database interactions and does not set any column size constraints. By the way, your Python version 3.9 is a bit dated, so I wonder if you are using an up-to-date version of SQLAlchemy?

**Comment by justinkumpe** [comment]
I believe not setting column size constraints is the problem. With MariaDB, if you create a column of type varchar then you must set it's length. I have updated to python 3.11 and still have same issue.

**Comment by rseichter** [comment]
I might have worded my response better, so let me try again: It should not be necessary for Fangfrisch to specify string length constraints. The SQLAlchemy [datatype documentation](https://docs.sqlalchemy.org/en/20/core/type_basics.html#the-camelcase-datatypes) states:

> The “CamelCase” types are to the greatest degree possible database agnostic, meaning they can all be used on any database backend where they will behave in such a way as appropriate to that backend in order to produce the desired behavior.

The documentation even specifically cites `String` as an example. While Fangfrisch _could_ specify a column length, it is not required, meaning that SQLAlchemy needs to be able to cope with the absence. I understand that this doesn't help you right now, but this is an issue the SQLAlchemy folks should look into. Fangfrisch knows nothing about database dialects, nor does it need to.

For the time being, I suggest you use a SQLite database as an alternative. Fangfrisch writes very little data, and infrequently. There is no need for transactional safety or other fancy measures. MariaDB is very much overkill here.

**Comment by justinkumpe** [comment]
That makes since. I found the .sql file and manually created the DBs and it is working now. I know MariaDB is overkill but I am already using MariaDB for other systems (including the mail server) so it was just easier for me to use what I already have.

**Comment by rseichter** [comment]
Good to hear you found a solution. Perhaps you would consider reporting the issue to the SQLAlchemy team anyway? I would have to set up a testing environment first in order to reproduce the problem.

Not that I am trying to weasel out of this, it is just a matter of separating responsibilities between Fangfrisch and SQLAlchemy. If I wanted to deal with database specifics, I would not have chosen SQLAlchemy as an abstraction layer.

**Comment by justinkumpe** [comment]
Yes, I absolutely intend to raise an issue with SQLAlchemy.
